rehash-openssl/src/lib.rs

```diff
@@ -0,0 +1,115 @@
+use std::{borrow::Cow, collections::HashMap};
+
+use pyo3::prelude::*;
+
+#[pyclass(module = "rehash_openssl", name = "sha256")]
+struct Sha256 {
+    hasher: openssl::sha::Sha256,
+}
+
+#[pymethods]
+impl Sha256 {
+    #[new]
+    fn new() -> Self {
+        Sha256 {
+            hasher: openssl::sha::Sha256::new(),
+        }
+    }
+
+    fn update(&mut self, data: &[u8]) {
+        self.hasher.update(data);
+    }
+
+    fn digest(&mut self) -> Cow<[u8]> {
+        let result = self.hasher.clone().finish();
+        Cow::Owned(result.to_vec())
+    }
+
+    fn hexdigest(&mut self) -> String {
+        let result = self.hasher.clone().finish();
+        hex::encode(result)
+    }
+
+    fn __getstate__(&self, py: Python) -> PyResult<HashMap<String, PyObject>> {
+        let state: &[u8] = unsafe {
+            std::slice::from_raw_parts(
+                &self.hasher as *const openssl::sha::Sha256 as *const u8,
+                std::mem::size_of::<openssl::sha::Sha256>(),
+            )
+        };
+        // return a dict that is compatible with rehash
+        let mut dict: HashMap<String, PyObject> = HashMap::new();
+        dict.insert("name".to_string(), "sha256".into_py(py));
+        dict.insert("md_data".to_string(), state.into_py(py));
+        Ok(dict)
+    }
+
+    fn __setstate__(&mut self, py: Python, state: HashMap<String, PyObject>) -> PyResult<()> {
```

Comment:
After talking with some rustaceans, it sounds like this could eventually use a pure-rust implementation that officially supports a resumption API. With that in mind, it might make sense to start adding a key that indicates the format of the serialized state to enable a seamless upgrade in the future.